### PR TITLE
Wii SD Card Folder Syncing

### PR DESCRIPTION
**Pull request by AdmiralCurtiss**
Rebased and modified #10525. This now behaves like the melonDS variant (I think) where it will, if enabled, sync a folder into an SD Card image on emulation start, and then sync it back from the image to the folder on emulation end.

The default folder that syncs with the image is at `[userdir]/Load/WiiSDSync`. **Be careful** as this will overwrite the contents of your `sd.raw` if you already have one!

As-is, this is pretty barebones but works. Additional features that could be nice and can be part of either this PR or a follow up include manual pack/unpack buttons, only syncing files that have changed (right now it just packs or unpacks the entire thing every time), some safety checks, and probably more. Maybe we should also take this opportunity to change the default path of the `sd.raw` to not be inside the Wii NAND...

I have not extensively tested this -- in fact most of the code is just taken as-is from #10525 -- so it's entirely possible this has issues, but from a little test it worked as expected.

### CODE REVIEW

**Comment by iwubcode** [comment]
> Be careful as this will overwrite the contents of your sd.raw if you already have one!

Doesn't have to be this PR but it'd be nice if the user clicks that sync checkbox and Dolphin notices there's a `sd.raw` that the files get unpacked to the folder.  Likewise if it's unchecked but "insert SD card" is checked, we could create the `sd.raw` from the folder.

Outside of that, this sounds good.  Haven't tested or looked at any code yet.

File: CMakeLists.txt [hunk 1]
```diff
@@ -871,6 +871,8 @@ include_directories(Externals/picojson)
 
 add_subdirectory(Externals/rangeset)
 
+add_subdirectory(Externals/FatFs)
```
Comment (shuffle2) [inline CMakeLists.txt:874]: missing the cmake file in the directory, appears to be causing other PRs to fail, which is weird.

**Comment by MayImilae** [comment]
Tested on macOS 12 and it works! And it's very convenient, super nice.

**Comment by JaredWhiteOne** [comment]
Tested on Windows 10 and works as expected. A little slow for larger SD cards, as expected really, but on second load there must have been some caching that occurred because it was dramatically faster.

**Comment by JMC47** [comment]
Does this crate deterministic SD cards - would it work for netplay?

**Comment by AdmiralCurtiss** [comment]
We probably should also use dummy timestamps and such in the deterministic case, now that I think about it.

**Comment by JMC47** [comment]
Yeah, that was the issue on old netplay - even if you put together the same exact SD card, unless you shared it directly with the other person, the timestamps wouldn't line up.

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild

**Comment by JaredWhiteOne** [comment]
@AdmiralCurtiss Does this sync the whole sd card image, or just whatever is different between the raw file and the folder?

**Comment by AdmiralCurtiss** [comment]
The entire thing. Trying to detect and process differences only could be added but for a first version this seems plenty already.

**Comment by JMC47** [comment]
Moving the SD card outside of the Wii NAND is correct.  I don't think anyone will argue with that.

**Comment by JosJuice** [comment]
Yeah, I agree with moving it outside of the Wii directory. I don't know about putting it in Load, since that's normally for stuff that Dolphin doesn't write to... But I'm not sure what a better place would be.

**Comment by JaredWhiteOne** [comment]
Maybe a new directory in the User folder?

**Comment by MayImilae** [comment]
Remember how I said this PR seemed a little crashy? Testing further, I found a way to reproduce it reliably. It's pretty simple. Just start the Wii Menu through the menu bar, and click A to get past the health screen and into the Wii Menu itself (which is necessary, it doesn't crash otherwise). Once the Wii Menu is up, stop emulation, wait for it to finish stopping, then open the Wii Menu the exact same way. This should be the crash. This does not occur on master.

**Comment by AdmiralCurtiss** [comment]
Cannot reproduce. Stack trace please, ideally from a debug build?

**Comment by Pokechu22** [comment]
I tested this briefly (I wanted to use some homebrew, and setting up an SD card is a pain otherwise). "Convert Folder to File now" seems to work fine.

I was a bit confused initially because I had assumed that "Automatically sync with Folder on emulation start and end" would default to checked, but it doesn't (which is probably for the better, as otherwise it'd default to replacing the SD card with an empty folder) - the relevant options are in Config &rarr; Wii. For my purposes, "Convert Folder to File now" was what I wanted though.

**Comment by JMC47** [comment]
Everything works as expected on my standard SD card.  But the one I converted from my actual Wii doesn't work.  It loads in Dolphin fine, but it doesn't convert to folder properly.

The loadtimes are pretty excessive when not using an m2 drive, which can be a bit annoying on 2GB cards.  at 128MB default card isn't that bad.

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Pull request merged**